src/Database/Kioku/Internal/TrieIndex.hs

```diff
@@ -23,22 +24,37 @@ import            System.IO
 import            Database.Kioku.Internal.Buffer
 import            Database.Kioku.Memorizable
 
-
 trieLookup :: BS.ByteString -> TrieIndex -> [Int]
 trieLookup key = maybe [] trieRootElems . lookupSubtrie key
 
 trieMatch :: BS.ByteString -> TrieIndex -> [Int]
 trieMatch prefix = maybe [] trieElems . lookupSubtrie prefix
 
+trieAllHitsAlong :: BS.ByteString -> TrieIndex -> [Int]
+trieAllHitsAlong "" _ = []
+trieAllHitsAlong key trie =
+  let (p,ath) = BS.splitAt 1 key
+  in concat $ go p ath trie
+    where
+      go prefix remaining trie' =
+        let (p', ath') = BS.splitAt 1 remaining
+        in case lookupSubtrie prefix trie' of
+             Nothing -> if BS.null remaining
+                          then []
+                          else go p' ath' trie'
+             Just subtrie -> case trieRootElems subtrie of
+                               []    -> go p' ath' subtrie
+                               elems -> elems : go p' ath' subtrie
```

Comment:
I would extract the comment `g p' ath' subtrie` here as a name above with the `let (p', ath')`. It took my quite a while to figure out what was going on here. For quite a bit I thought we were double-splitting.